### PR TITLE
Opf integration with dc results

### PR DESCRIPTION
**Pull request by albanpuech**
- added OPF from cleaned version of #39 
    - Qg penalty loss computed inly on buses with violation
    - split from existing split file (for handling opfdata)

- OPF evaluation
    - Qg violation on PV + REF bus
    - Fixed issue in angle conversion for branch angle difference constraint
    - added note about aggregation of the metrics for OPF
 
- OPF DC baseline
    - added dc baseline for OPF for datakit data
    - added support for predict

- docs and config file
    - added config file for opfdata
    - removed all legacy docs and updated installation, readme, quickstart and cli
    
- compilation:
     - fixed loading of weight of compiled models
     
- MPS compatibility/ lsf support/ multigpu training

### CODE REVIEW

File: docs/quick_start/quick_start.md [hunk 1]
```diff
@@ -8,10 +8,11 @@ gridfm_graphkit <command> [OPTIONS]
...
 
-* `train` – Train a new model from scratch
-* `finetune` – Fine-tune an existing pre-trained model
-* `evaluate` – Evaluate model performance on a dataset
-* `predict` – Run inference and save predictions
+* `train` - Train a new model from scratch
+* `finetune` - Fine-tune an existing pre-trained model
+* `evaluate` - Evaluate model performance on a dataset
+* `predict` - Run inference and save predictions
+* `benchmark` - Benchmark train-dataloader iteration speed
```
Comment (albanpuech) [inline docs/quick_start/quick_start.md:15]: @romeokienzler do we still need the benchmark?

File: docs/quick_start/quick_start.md [hunk 2]
```diff
@@ -105,16 +132,44 @@ gridfm_graphkit predict --config path/to/config.yaml --model_path path/to/model.
...
+| `--bfloat16` | `flag` | Cast model to `torch.bfloat16` (`model.to(torch.bfloat16)`). | `False` |
+| `--tf32` | `flag` | Enable TF32 on Ampere+ GPUs via `torch.set_float32_matmul_precision("high")`. | `False` |
+| `--profiler` | `str` | Enable Lightning profiler (`simple`, `advanced`, `pytorch`). | `None` |
+
+---
+
+## Benchmarking Dataloader Throughput
+
+```bash
+gridfm_graphkit benchmark --config path/to/config.yaml
```
Comment (albanpuech) [inline docs/quick_start/quick_start.md:159]: @romeokienzler this is OK?

File: gridfm_graphkit/datasets/normalizers.py [hunk 3]
```diff
@@ -228,8 +228,8 @@ def transform(self, data: HeteroData):
         data.edge_attr_dict[("bus", "connects", "bus")][:, ANG_MIN] *= torch.pi / 180.0
         data.edge_attr_dict[("bus", "connects", "bus")][:, ANG_MAX] *= torch.pi / 180.0
         data.edge_attr_dict[("bus", "connects", "bus")][:, RATE_A] /= self.baseMVA
-        data.baseMVA = self.baseMVA
-        data.is_normalized = True
+        data.baseMVA = torch.tensor(self.baseMVA, dtype=data.x_dict["bus"].dtype) # # needs to be float32 for MPS
+        data.is_normalized = torch.tensor(True, dtype=torch.bool) # needs to be bool for MPS
```
Comment (albanpuech) [inline gridfm_graphkit/datasets/normalizers.py:232]: let me know if you see any problem with this? @romeokienzler 
regression test passed

File: gridfm_graphkit/tasks/base_task.py [hunk 4]
```diff
@@ -20,6 +20,20 @@ def __init__(self, args, data_normalizers):
...
+        get collated as float64 even when model inputs are float32, so coerce them
+        first and then delegate to Lightning's standard device transfer.
+        """
+        if getattr(device, "type", None) == "mps" and hasattr(batch, "stores"):
+            for store in batch.stores:
+                for key, val in store.items():
+                    if isinstance(val, torch.Tensor) and val.dtype == torch.float64:
+                        store[key] = val.to(torch.float32)
+        return super().transfer_batch_to_device(batch, device, dataloader_idx)
+
```
Comment (albanpuech) [inline gridfm_graphkit/tasks/base_task.py:36]: do you see any issue with this @romeokienzler ? 
Thanks to this, the model now works on mps

File: gridfm_graphkit/tasks/opf_task.py [hunk 5]
```diff
@@ -293,8 +302,25 @@ def test_step(self, batch, batch_idx, dataloader_idx=0):
...
+                for rank_data in gathered:
+                    for dl_idx, batches in rank_data.items():
+                        merged[dl_idx].extend(batches)
+                self.test_outputs = merged
+
+        # Only rank 0 proceeds with logging, CSV writing, and plotting
+        if dist.is_available() and dist.is_initialized() and dist.get_rank() != 0:
+            self.test_outputs.clear()
+            return
+
```
Comment (albanpuech) [inline gridfm_graphkit/tasks/opf_task.py:323]: in the end i kept multi gpu evaluation bc it was too long to change @romeokienzler

File: gridfm_graphkit/tasks/pf_task.py [hunk 6]
```diff
@@ -241,6 +241,7 @@ def on_test_end(self):
 
         # Only rank 0 proceeds with logging, CSV writing, and plotting
         if dist.is_available() and dist.is_initialized() and dist.get_rank() != 0:
+            self.test_outputs.clear() # clear the test outputs for other ranks
```
Comment (albanpuech) [inline gridfm_graphkit/tasks/pf_task.py:244]: apparently we needed this @romeokienzler

File: gridfm_graphkit/tasks/utils.py [hunk 7]
```diff
@@ -7,10 +7,25 @@
...
+        abs_residual_cpu = abs_residual.cpu()
+        batch_masked_cpu = batch_masked.cpu()
+        mean_res = scatter_mean(abs_residual_cpu, batch_masked_cpu, dim=0).to(
+            abs_residual.device,
+        )
+        max_res, _ = scatter_max(abs_residual_cpu, batch_masked_cpu, dim=0)
+        max_res = max_res.to(abs_residual.device)
+    else:
+        mean_res = scatter_mean(abs_residual, batch_masked, dim=0)
+        max_res, _ = scatter_max(abs_residual, batch_masked, dim=0)
```
Comment (albanpuech) [inline gridfm_graphkit/tasks/utils.py:28]: that was the trick for mps ;)

File: gridfm_graphkit/training/callbacks.py [hunk 8]
```diff
@@ -52,6 +53,15 @@ def __init__(
...
 
+    @staticmethod
+    def _canonical_state_dict(pl_module):
+        """Return a state dict with compile wrappers removed from key names."""
+        state_dict = pl_module.state_dict()
+        return {
+            key.replace("model._orig_mod.", "model."): value
+            for key, value in state_dict.items()
+        }
+
```
Comment (albanpuech) [inline gridfm_graphkit/training/callbacks.py:64]: to handle state dicts from compiled models @romeokienzler

File: gridfm_graphkit/training/loss.py [hunk 9]
```diff
@@ -322,3 +338,57 @@ def forward(
...
+        if Qg_under!=Qg_under: # replacing nan with 0 
+            Qg_under = 0.0
+
+        penalty_loss = Qg_over + Qg_under            
+        loss += penalty_loss
+
+        try:
+            output = {"loss": loss, "Qg Violation Penalty loss": loss.detach()}
+        except:
+            output = {"loss": loss, "Qg Violation Penalty loss": loss}
```
Comment (albanpuech) [inline gridfm_graphkit/training/loss.py:391]: i don't like this so much, will change it later

File: gridfm_graphkit/__main__.py [hunk 10]
```diff
@@ -3,7 +3,64 @@
...
 
+import subprocess
+import os
+
+def is_lsf():
+    return (
+        os.environ.get("LSB_JOBID") is not None
+        and os.environ.get("LSB_MCPU_HOSTS") is not None
+        and "LSF_ENVDIR" in os.environ  # strong LSF indicator
+    )
```
Comment (albanpuech) [inline gridfm_graphkit/__main__.py:14]: @romeokienzler to detect if we are on ccc and need the fix_infiniband and set_env functions to run

File: gridfm_graphkit/cli.py [hunk 11]
```diff
@@ -15,14 +15,26 @@
...
+def _normalize_loaded_state_dict_keys(state_dict: dict[str, torch.Tensor]) -> dict[str, torch.Tensor]:
+    """Map legacy torch.compile checkpoint keys to the canonical model namespace."""
+    has_compiled_prefix = any(key.startswith("model._orig_mod.") for key in state_dict)
+    if not has_compiled_prefix:
+        return state_dict
+    return {
+        key.replace("model._orig_mod.", "model."): value
+        for key, value in state_dict.items()
+    }
+
```
Comment (albanpuech) [inline gridfm_graphkit/cli.py:36]: handle torch compile checkpoint keys

File: gridfm_graphkit/cli.py [hunk 12]
```diff
@@ -305,19 +337,27 @@ def main_cli(args):
...
+                    output_dir,
+                    f"{grid_name}{suffix}_predictions.parquet",
+                )
+                df.to_parquet(out_path, index=False)
+                print(f"Saved {table_name} predictions to {out_path}")
+        else:
+            df = _predictions_to_dataframe(predictions)
+            out_path = os.path.join(output_dir, f"{grid_name}_predictions.parquet")
+            df.to_parquet(out_path, index=False)
+            print(f"Saved predictions to {out_path}")
```
Comment (albanpuech) [inline gridfm_graphkit/cli.py:363]: i dont think this works with multi gpu so i made sure the trainer is created with one gpu only

File: gridfm_graphkit/datasets/hetero_powergrid_datamodule.py [hunk 13]
```diff
@@ -167,53 +183,93 @@ def setup(self, stage: str):
 
             # Create a subset
             all_indices = list(range(len(dataset)))
-            # Random seed set before every shuffle for reproducibility in case the power grid datasets are analyzed in a different order
-            random.seed(self.args.seed)
-            random.shuffle(all_indices)
-            subset_indices = all_indices[:num_scenarios]
 
-            # load_scenario for each scenario in the subset
-            load_scenarios = dataset.load_scenarios[subset_indices]
```
Comment (romeokienzler) [inline gridfm_graphkit/datasets/hetero_powergrid_datamodule.py:187]: @albanpuech so we don't need this anymore?

File: gridfm_graphkit/datasets/hetero_powergrid_datamodule.py [hunk 14]
```diff
@@ -386,6 +442,7 @@ def _dataloader_kwargs(self):
         return kwargs
 
     def train_dataloader(self):
+        print("creating train dataloader for rank ", dist.get_rank() if dist.is_available() and dist.is_initialized() else "not distributed")
```
Comment (romeokienzler) [inline gridfm_graphkit/datasets/hetero_powergrid_datamodule.py:445]: @albanpuech just cosmetics, but it might print

creating train dataloader for rank not distributed

File: gridfm_graphkit/datasets/hetero_powergrid_datamodule.py (hunk 13 above)
Comment (albanpuech) [reply gridfm_graphkit/datasets/hetero_powergrid_datamodule.py:187]: now it is done here: https://github.com/gridfm/gridfm-graphkit/pull/54/changes/BASE..b016e7e6e40c02905087ebc8716f14c6be939739#diff-6adb4ccf8a9ab80ddbf51f2133b6a30995053e5e0eabc97a0409620148f64034R230

File: gridfm_graphkit/datasets/hetero_powergrid_datamodule.py (hunk 14 above)
Comment (albanpuech) [reply gridfm_graphkit/datasets/hetero_powergrid_datamodule.py:445]: ok for now but agreed

File: gridfm_graphkit/datasets/normalizers.py (hunk 3 above)
Comment (romeokienzler) [reply gridfm_graphkit/datasets/normalizers.py:232]: @albanpuech looks very good to me, only gotcha I've read about:

if x is True does not work anymore and needs to be replaced by if x or x.item()

File: gridfm_graphkit/datasets/normalizers.py [hunk 15]
```diff
@@ -299,7 +299,7 @@ def inverse_transform(self, data: HeteroData):
         data.edge_attr_dict[("bus", "connects", "bus")][:, ANG_MAX] *= 180.0 / torch.pi
 
         data.edge_attr_dict[("bus", "connects", "bus")][:, RATE_A] *= self.baseMVA
-        data.is_normalized = False
+        data.is_normalized = torch.tensor(False, dtype=torch.bool) # needs to be bool for MPS
```
Comment (romeokienzler) [inline gridfm_graphkit/datasets/normalizers.py:302]: @albanpuech cool, well spottet :)

File: gridfm_graphkit/datasets/normalizers.py (hunk 3 above)
Comment (albanpuech) [reply gridfm_graphkit/datasets/normalizers.py:232]: we dont need that so perfect

File: gridfm_graphkit/tasks/base_task.py (hunk 4 above)
Comment (romeokienzler) [reply gridfm_graphkit/tasks/base_task.py:36]: The only recommendation I got was adding this:

```
# Also check top-level attributes not in 'stores'
for key, val in batch.__dict__.items():
    if isinstance(val, torch.Tensor) and val.dtype == torch.float64:
        setattr(batch, key, val.to(torch.float32))
```

File: gridfm_graphkit/tasks/opf_task.py (hunk 5 above)
Comment (romeokienzler) [reply gridfm_graphkit/tasks/opf_task.py:323]: ok, now worries, we can always do that later

File: gridfm_graphkit/tasks/pf_task.py (hunk 6 above)
Comment (romeokienzler) [reply gridfm_graphkit/tasks/pf_task.py:244]: congrats for finding this out!

File: gridfm_graphkit/tasks/utils.py (hunk 7 above)
Comment (romeokienzler) [reply gridfm_graphkit/tasks/utils.py:28]: @albanpuech Does this function get called during training_step or only during validation_step? If it's only for validation/logging, your current implementation is perfectly safe. If it's for training, you might want to look into whether you can avoid the max operation or use a native PyTorch equivalent that might have better MPS support in your specific version.

File: gridfm_graphkit/training/callbacks.py (hunk 8 above)
Comment (romeokienzler) [reply gridfm_graphkit/training/callbacks.py:64]: @albanpuech 
this might be a bit more robust/generic?

`return {key.replace("_orig_mod.", ""): value for key, value in state_dict.items()}`

File: gridfm_graphkit/training/loss.py (hunk 9 above)
Comment (romeokienzler) [reply gridfm_graphkit/training/loss.py:391]: ok, maybe add a #todo to just have a chance to ever find it?

File: gridfm_graphkit/__main__.py (hunk 10 above)
Comment (romeokienzler) [reply gridfm_graphkit/__main__.py:14]: question is if we wanna detect ccc or lsf? I guess ccc, so we could just query the hostname like this:

```
import socket

def is_lsf():
    """Returns True only if 'cccxc' is present in the system hostname."""
    return "cccxc" in socket.gethostname().lower()
```

File: gridfm_graphkit/cli.py (hunk 12 above)
Comment (romeokienzler) [reply gridfm_graphkit/cli.py:363]: fine, you could also have used, jfyi
> if self.trainer.is_global_zero:

**Review by romeokienzler — APPROVED** [review]
@albanpuech lgtm

**Comment by romeokienzler** [comment]
@albanpuech ok for me to override the pre-commit run as we agreed to run it at a lager stage
